lib/torrent.js

```diff
@@ -638,24 +704,110 @@ export default class Torrent extends EventEmitter {
     })
   }
 
-  _verifyPieces (cb) {
-    parallelLimit(this.pieces.map((piece, index) => cb => {
+  /**
+   * Callback called after a piece is verified
+   * @callback verifyPieceCallback
+   * @param {Error=} error
+   * @param {boolean=} isValid
+   */
+  /**
+   * Verify a single piece using hashing
+   * @param index
+   * @param {verifyPieceCallback} cb
+   * @private
+   */
+  _verifyPiece (index, cb) {
+    if (this.destroyed) return cb(new Error('torrent is destroyed'))
+
+    const getOpts = {}
+    // Specify length for the last piece in case it is zero-padded
+    if (index === this.pieces.length - 1) {
+      getOpts.length = this.lastPieceLength
+    }
+    this.store.get(index, getOpts, async (err, buf) => {
       if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-      const getOpts = {}
-      // Specify length for the last piece in case it is zero-padded
-      if (index === this.pieces.length - 1) {
-        getOpts.length = this.lastPieceLength
-      }
-      this.store.get(index, getOpts, async (err, buf) => {
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      if (err) return queueMicrotask(() => cb(null, false)) // ignore error
 
-        if (err) return queueMicrotask(() => cb(null)) // ignore error
+      const hex = await hash(buf, 'hex')
+      if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-        const hex = await hash(buf, 'hex')
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      cb(null, hex === this._hashes[index])
+    })
+  }
+
+  /**
+   * Verify pieces using bitfield, in case of in-congruences it will re-verify the file using hashing
+   * @param {callbackWithError} cb
+   * @private
+   */
+  _verifyPiecesUsingBitfield (cb) {
+    const piecesToCheck = new Set()
+    const piecesToFilesMap = new Map()
+    this.files.forEach(file => {
+      const previouslyVerifiedFilePieces = []
+      for (let i = file._startPiece; i <= file._endPiece; ++i) {
+        if (this.bitfield.get(i)) {
+          previouslyVerifiedFilePieces.push(i)
```

Comment:
wow i hate this, actually, i despise this, this is way too much, i have 0 clue what even is happening here at a glance, idk why you wrote it this way

Reply:
I liked the 2nd piece approach better, because the only pieces that can be overlapping is first or last, 2nd means its not first, if its last, oh well you cant do anything about it anyways, but at least we know its not first

Reply:
The problem is that we don't know exactly what is downloaded in advance so we can't really check the second piece.

Imagine the file is split in 10 pieces, and the download mode is rarest first, and it fully downloaded 3rd piece and 5th piece. The 2nd piece is not downloaded so we can't use it for verification, and the second of the downloaded pieces is in reality the 5th piece.

The way I wrote it:

1. Find what pieces are downloaded for that file ( in our example it would be 3rd and 5th )
2. Find the first piece that is not overlapping ( it would be the 3rd aka 1st )
3. Fallback case, if only one piece is downloaded use that one even if overlaps

Reply:
why....? the if the 2nd piece is in the preloaded bitfield, then it has been downloaded, if its not then it will check if the file exists using 1st piece

Reply:
sorry, my code doesnt check the 2nd piece explicity, just a `2nd piece for the file` we dont know which piece in the file it is, it could be the 14th piece, because the file only has 10th piece and 14th piece downloaded, that's fine too

Reply:
I'm thinking of the edge casa when we have a single piece. 

The other solution would be to collect these pieces and if we have more than one we use the second one, in the case we have only one piece we use the first one.

I used a reverse map to know if the piece is multiple files for the edge case when the file has only one single piece downloaded and it overlaps ( it could be because we downloaded only the first piece or the file is smaller than one piece ( I imagine that in a torrent with a 4K movie and various subtitles, the subtitles files could fit all in one file ))

Reply:
then it never reaches the 2nd part of the check which replaces the first piece with the second?

Reply:
It should be able to reach the last condition in the case where there's only one piece or all the piece overlap ( first and last piece downloaded )

Reply:
yes, what i suggest, adds the first piece to check, then if the 2nd piece exists on the next iteration, remvoes the first piece and adds the 2nd piece in its place

Reply:
what you got *could* be better? i'm not sure because the code is so convoluted i can't really see what it does at a glance

Reply:
I'm thinking of a way to make the code very understandable


```js
// First step, optimistically mark what is verified and what is not by blindly trusting the bitfield
    this.files.forEach(file => {
      for (let i = file._startPiece; i <= file._endPiece; ++i) {
        if (this.bitfield.get(i)) {
          this.emit('verified', i)
        } else {
          this._markUnverified(i)
          this._debug('piece invalid %s', i)
        }
      }
    })
    // Second step, we construct support structures to choose what pieces should be slowly verified using hashing
    const piecesToFilesMap = new Map()
    // We are going to use this to discover what pieces are unique to each file
    const downloadedPiecesByFile = this.files.map(file => {
      const downloadedPieces = []
      for (let i = file._startPiece; i <= file._endPiece; ++i) {
        if (this.bitfield.get(i)) {
          if (!piecesToFilesMap.has(i)) {
            piecesToFilesMap.set(i, [])
          }
          piecesToFilesMap.get(i).push(file)
          downloadedPieces.push(i)
        }
      }
      return downloadedPieces
    })
    // Third step, using the support structures we are going to choose what pieces to verify
    const uniquePiecesByFile = downloadedPiecesByFile.map((downloadedPieces) => {
      // If there's only one piece we don't care if it's unique
      if (downloadedPieces.length === 1) {
        return downloadedPieces
      }
      // piecesToFilesMap tells us how many files share that piece, if it's only one than it's unique to that file
      return downloadedPieces.filter(piece => piecesToFilesMap.get(piece).length === 1)
    })
    const piecesToCheck = new Set(uniquePiecesByFile.map(pieces => pieces[0]))
```

I believe this is very robust and at the same time should be understandable

Reply:
that's way too much code, way too many iterations for something way too simple, i'm telling you, simply
``js
for (const file of this.files) {
  for (let i = file._startPiece; i <= file._endPiece; ++i) {
    let checkFile = 2
    if (checkFile) {
      piecesToCheck.add(i)
      if (checkFile-- === 1) {
        piecesToCheck.delete(i - 1)
      }
    }
  }
}
```
will do most of the same work, while being loads faster and easier to understand

Reply:
> piecesToCheck.delete(i - 1)

We have no insurance it was the piece immediately before, so we need another support variable for it

Reply:
This should do the trick:

```js
const piecesToCheck = new Set()
    // First step, optimistically mark what is verified and what is not by blindly trusting the bitfield
    this.files.forEach(file => {
      let checkFile = 2
      let pieceToCheckForThisFile = null
      for (let i = file._startPiece; i <= file._endPiece; ++i) {
        if (this.bitfield.get(i)) {
          if (checkFile) {
            pieceToCheckForThisFile = i
            checkFile--
          }
          this.emit('verified', i)
        } else {
          this._markUnverified(i)
          this._debug('piece invalid %s', i)
        }
      }
      piecesToCheck.add(pieceToCheckForThisFile)
    })
```